### PR TITLE
feat:[SSCA-605]: Added Changes in DB Schema and Orchestration Service

### PR DESCRIPTION
**Pull request by arpit-harness**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/52821)

### CODE REVIEW

**Comment by arpit-harness** [comment]
trigger uts

**Comment by arpit-harness** [comment]
trigger uts

File: ssca-manager/service/src/main/java/io/harness/ssca/entities/ArtifactEntity.java [hunk 1]
```diff
@@ -66,4 +74,6 @@ public static class Sbom {
     @Field("sbomformat") String sbomFormat;
     @Field("sbomversion") String sbomVersion;
   }
+
+  public enum Activity { DEPLOYED, NON_DEPLOYED }
```
Comment (inderschera) [inline ssca-manager/service/src/main/java/io/harness/ssca/entities/ArtifactEntity.java:78]: Create separate class for this

File: ssca-manager/service/src/main/java/io/harness/ssca/services/OrchestrationStepServiceImpl.java [hunk 2]
```diff
@@ -64,16 +64,24 @@ public String processSBOM(String accountId, String orgIdentifier, String project
...
 
     SettingsDTO settingsDTO =
         getSettingsDTO(accountId, orgIdentifier, projectIdentifier, sbomProcessRequestBody, artifactEntity);
 
     Normalizer normalizer = normalizerRegistry.getNormalizer(settingsDTO.getFormat()).get();
     List<NormalizedSBOMComponentEntity> sbomEntityList = normalizer.normaliseSBOM(sbomDTO, settingsDTO);
 
+    artifactEntity.setComponentsCount(sbomEntityList.stream().count());
+
+    artifactRepository.saveAndUpdateInvalidFlag(artifactEntity);
```
Comment (inderschera) [inline ssca-manager/service/src/main/java/io/harness/ssca/services/OrchestrationStepServiceImpl.java:84]: We should do this in transaction

File: ssca-manager/service/src/main/java/io/harness/repositories/ArtifactRepositoryCustomImpl.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+import org.springframework.transaction.annotation.Transactional;
+
+@OwnedBy(SSCA)
+@AllArgsConstructor(access = AccessLevel.PROTECTED, onConstructor = @__({ @Inject }))
+public class ArtifactRepositoryCustomImpl implements ArtifactRepositoryCustom {
+  private final MongoTemplate mongoTemplate;
+
+  @Override
+  @Transactional
+  public void saveAndUpdateInvalidFlag(ArtifactEntity artifact) {
```
Comment (inderschera) [inline ssca-manager/service/src/main/java/io/harness/repositories/ArtifactRepositoryCustomImpl.java:32]: We don't need to do update all older artifacts. Just the current latest.
We will do migration to take database to such state

File: ssca-manager/service/src/main/java/io/harness/ssca/services/OrchestrationStepServiceImpl.java (hunk 2 above)
Comment (arpit-harness) [reply ssca-manager/service/src/main/java/io/harness/ssca/services/OrchestrationStepServiceImpl.java:84]: it is transactional.

File: ssca-manager/service/src/main/java/io/harness/repositories/ArtifactRepositoryCustomImpl.java (hunk 3 above)
Comment (arpit-harness) [reply ssca-manager/service/src/main/java/io/harness/repositories/ArtifactRepositoryCustomImpl.java:32]: Added changes to update only the valid artifacts.

File: ssca-manager/service/src/main/java/io/harness/ssca/entities/ArtifactEntity.java (hunk 1 above)
Comment (arpit-harness) [reply ssca-manager/service/src/main/java/io/harness/ssca/entities/ArtifactEntity.java:78]: Done

**Comment by arpit-harness** [comment]
trigger uts

File: ssca-manager/service/src/main/java/io/harness/repositories/ArtifactRepositoryCustomImpl.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    long currentTime = Instant.now().toEpochMilli();
+    Criteria criteria = Criteria.where("artifactid")
```
Comment (reetika04) [inline ssca-manager/service/src/main/java/io/harness/repositories/ArtifactRepositoryCustomImpl.java:34]: Please use keys "ArtifactEntityKeys.artifactId". Ideally this logic should be in the service. Please add a new method to save and get using the criteria in the repository. Move this method to the service.

**Review by deepharness — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52821)* status: 10 file(s) to review. 1 change requests. 3. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`ssca-manager/service/src/main/java/io/harness/repositories/ArtifactRepositoryCustomImpl.java` line 46 at r4](https://reviewable.io/reviews/harness/harness-core/52821#-NedCQmz-WxPAuavu1rG:-NedCQmz-WxPAuavu1rH:bdq7w1x) ([raw file](https://github.com/harness/harness-core/blob/02636f8d0042ba659150110c75a9f36c938dd327/ssca-manager/service/src/main/java/io/harness/repositories/ArtifactRepositoryCustomImpl.java#L46)):*
> ```Java
>     Update update = new Update();
>     update.set(ArtifactEntityKeys.invalid, true);
>     update.set(ArtifactEntityKeys.lastUpdatedAt, currentTime);
> ```

we don't want to update the lastUpdatedAt for all the matching tag , I guess it should be updated only when we are adding or updating the field other than invalid 
___
*[`ssca-manager/service/src/main/java/io/harness/ssca/services/ArtifactServiceImpl.java` line 84 at r4](https://reviewable.io/reviews/harness/harness-core/52821#-NedDnRqCSeEtdJUXiCi:-NedDnRqCSeEtdJUXiCj:b-xayipu) ([raw file](https://github.com/harness/harness-core/blob/02636f8d0042ba659150110c75a9f36c938dd327/ssca-manager/service/src/main/java/io/harness/ssca/services/ArtifactServiceImpl.java#L84)):*
> ```Java
>   public void saveArtifactAndInvalidateOldArtifact(ArtifactEntity artifact) {
>     long currentTime = artifact.getCreatedOn().toEpochMilli();
>     artifactRepository.invalidateOldArtifact(artifact, currentTime);
> ```

No need to update currentTime for old artifact

**Comment by arpit-harness** [comment]
trigger uts

**Comment by arpit-harness** [comment]
trigger uts

**Review by deepharness — APPROVED** [review]
> Reviewed 3 of 7 files at r1, 2 of 3 files at r3, 1 of 5 files at r4, 5 of 5 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52821)* status: 3. unresolved discussions.

**Pull request merged**